File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        char[] tArr = t.toCharArray();
        Map<Character, Integer> map = new HashMap<>();

        for (char c : tArr) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int count = 0;
        int minLen = Integer.MAX_VALUE;
        int start = 0;
        int windowLeft = 0;
        char[] sArr = s.toCharArray();
        for (int windowRight = 0; windowRight < sArr.length; windowRight++) {
            char c = sArr[windowRight];
            if (map.containsKey(c)) {
                map.put(c, map.get(c) - 1);
                if (map.get(c) >= 0) {
                    count++;
                }
            }
            
            while (count == t.length()) {
                if (windowRight - windowLeft + 1 < minLen) {
                    minLen = windowRight - windowLeft + 1;
                    start = windowLeft;
                }

                char leftChar = sArr[windowLeft];
                if (map.containsKey(leftChar)) {
                    map.put(leftChar, map.get(leftChar) + 1);
                    if (map.get(leftChar) > 0) {
                        count--;
                    }
                }

                windowLeft++;
            }
        }

        return minLen == Integer.MAX_VALUE ? "" : new String(sArr, start, minLen);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.length() == 0 || t.length() == 0) {
            return "";
        }

        unordered_map<char, int> map;
        for (char c : t) {
            map[c]++;
        }
        
        int windowLeft = 0;
        int matched = 0;
        int start = 0;
        int result = INT_MAX;
        for (int windowRight = 0; windowRight < s.length(); windowRight++) {
            char cur = s[windowRight];
            if (map.find(cur) != map.end()) {
                map[cur]--;
                if (map[cur] >= 0) {
                    matched++;
                }
            }

            while (matched == t.length()) {
                if (windowRight - windowLeft + 1 < result) {
                    result = windowRight - windowLeft + 1;
                    start = windowLeft;
                }
                
                char left = s[windowLeft];
                if (map.find(left) != map.end()) {
                    map[left]++;
                    if (map[left] > 0) {
                        matched--;
                    }
                }

                windowLeft++;
            }
        }

        return result == INT_MAX ? "" : s.substr(start, result);
    }
};